files-publisher/delta-cache.js

```diff
@@ -64,41 +64,104 @@ export default class DeltaCache {
     }
   }
 
+
+  async countDeltaFiles(service_config, since) {
+    console.log(`Retrieving count delta files since ${since}`);
+
+    const result = await query(`
+    ${service_config.prefixes}
+
+    SELECT (COUNT (distinct ?uuid) as ?count) WHERE {
+        ?s a nfo:FileDataObject ; 
+            mu:uuid ?uuid ;
+            nfo:fileName ?filename ;
+            dct:publisher <${service_config.publisherUri}> ;
+            dct:created ?created .
+        ?file nie:dataSource ?s .
+
+        FILTER (?created > "${since}"^^xsd:dateTime)
+     } 
+  `);
+
+    if (result.results.bindings?.length) {
+      return parseInt(result.results.bindings[0].count?.value || 0);
+    }
+    return 0;
+  }
+
   /**
    * Get all delta files produced since a given timestamp
    *
    * @param service_config the configuration to be used
    * @param since {string} ISO date time
    * @public
   */
-  async getDeltaFiles(service_config, since) {
+  async getDeltaFiles(service_config, since, page) {
+    const path = service_config.filesPath;
     console.log(`Retrieving delta files since ${since}`);
-
-    const result = await query(`
-    ${service_config.prefixes}
-
-    SELECT ?uuid ?filename ?created WHERE {
-      ?s a nfo:FileDataObject ;
-          mu:uuid ?uuid ;
-          nfo:fileName ?filename ;
-          dct:publisher <${service_config.publisherUri}> ;
-          dct:created ?created .
-      ?file nie:dataSource ?s .
-
-      FILTER (?created > "${since}"^^xsd:dateTime)
-    } ORDER BY ?created
-  `);
-
-    return result.results.bindings.map(b => {
+    const calculatePages = (totalCount, limit) => {
+      return Math.ceil(totalCount / limit);
+    };
+    const count = await this.countDeltaFiles(service_config, since);
+    console.log("found delta files count:", count);
+    const totalPages = calculatePages(count, DELTA_FILES_PAGINATION_MAX_PER_PAGE);
+    console.log("total pages:", totalPages);
+    const getPage = async (page, limit) => {
+      const offset = (page - 1) * limit;
+      const result = await query(`
+        ${service_config.prefixes}
+        SELECT ?uuid ?filename ?created WHERE {
+          SELECT distinct ?uuid ?filename ?created WHERE {
+            ?s a nfo:FileDataObject ;
+                mu:uuid ?uuid ;
+                nfo:fileName ?filename ;
+                dct:publisher <${service_config.publisherUri}> ;
+                dct:created ?created .
+            ?file nie:dataSource ?s .
+
+            FILTER (?created > "${since}"^^xsd:dateTime)
+          } ORDER BY ?uuid ?filename ?created
```

Comment:
Here you need to first order by `created` and then the two others no, to ensure that the order of execution will be kept?